lib/Sema/CSSimplify.cpp

```diff
@@ -7166,7 +7195,8 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     }
   }
 
-  if (kind >= ConstraintKind::Subtype) {
+  if (kind >= ConstraintKind::Subtype &&
+      type1->is<PackExpansionType>() == type2->is<PackExpansionType>()) {
```

Comment:
I think it would be safer if the block that handles `lhs->is<PackExpansionType>() != rhs->is<PackExpansionType>()` you added to `repairFailures` was moved right before `if (kind >= ConstraintKind::Conversion)` to make sure that we always check that expansions match on both sides before checking anything else.

Reply:
Good idea!